Added SoupRot

Masonjar13 opened the pull request:
Added a custom rotation algorithm I wrote. Not 100% sure I updated it properly, feel free to edit.

Masonjar13 committed:
Create SoupRot.ahk

Masonjar13 committed:
Add files via upload

Masonjar13 committed:
Update libcrypt.ahk

joedf commented on build/libcrypt.ahk:
```diff
@@ -1,1221 +1,1381 @@
-LC_Version := "0.0.24.03"
```
This should be updated once by simply running `make.ahk` once: it copies all scripts/functions into this one file `build/libcrypt.ahk`.

Masonjar13 replied:
I did use the make.ahk, and this was the output I received. Did I need to change the version file or something?

joedf replied:
Whoops, my bad. You're right, there is a file called "VERSION". I'd say increment it to "0.0.24.04" ?

joedf commented on examples/SoupRot.ahk:
```diff
@@ -0,0 +1,11 @@
+encStr:=soupRot.enc(str,iterationCnt,junk)
+decStr:=soupRot.dec(encStr,iterationCnt,junk)
+
+msgbox,,SoupRot,% encStr "`n`n" decStr
```
Maybe show the `str` before encoding, then the encoded string and finally the decoded string.
Perhaps something like:
```ahk
msgbox,,SoupRot, Original: %str%`nEncoded: %encStr%`nDecoded: %decStr%
```

Masonjar13 replied:
Done!

joedf commented on src/SoupRot.ahk:
```diff
@@ -0,0 +1,159 @@
+
+*/
+
+class soupRot {
```
Neat! first implementation that uses a `class`. All other implementations are single functions for simple in/out or enc/dec functions. That said, I don't have anything against this. I believe there's also an issue/work-item to make libcrypt into a class. For now, please prefix it with `LC_` for it to comply with all other functions, likely: `LC_soupRot`... to make it more "self-contained" or rather reduce loose names in global space -> having everything prefixed with `LC_`.

Masonjar13 replied:
Those dependencies were the reason for the class, actually. I work very modularly and rely heavily on my stdlib. It would definitely be better to have this entire library as a class (I certainly can work on that later), but I'll see what I can do about reducing the functions.

joedf commented on src/SoupRot.ahk:
```diff
@@ -0,0 +1,159 @@
+		}
+		return str
+	}
+	rand(lowerBound,upperBound){
```
Is it possible to move all of these utility functions into the enc/dec functions? Probably not ideal. If not, please prefix them with `LC_soupRot_UTILITYFUNCTIONNAME` or `__LC_UTILITYFUNCTIONNAME` or perhaps `__LC_soupRot_UTILITYFUNCTIONNAME` to make them specific. I'd probably go with this second option. That is, if a class is not used.
An example of this can be found here, you can prefix however you like:
https://github.com/ahkscript/libcrypt.ahk/blob/master/src/LHash.ahk

Masonjar13 replied:
I'll revisit this idea for further integration, but for now they've been "privatized."

joedf replied:
"privatized" neat, yes haha!

joedf commented on src/SoupRot.ahk:
```diff
@@ -0,0 +1,159 @@
+		if in is digit
+			return 1
+	}
+	b64e(byRef outData,byRef inData ){
```
We have [`LC_Base64_*`](https://github.com/ahkscript/libcrypt.ahk/blob/master/src/Base64.ahk) for this, which in turn also use `Crypt32.dll`'s `CryptStringToBinary`/`CryptBinaryToString`. Please use it if possible, if not please prefix them (if not using a class).
```ahk
LC_Base64_EncodeText(Text,Encoding="UTF-8")
LC_Base64_DecodeText(Text,Encoding="UTF-8")
LC_Base64_Encode(ByRef Out, ByRef In, InLen)
LC_Base64_Decode(ByRef Out, ByRef In)
```

Masonjar13 replied:
Done! Switched to using integrated Base64 functions.

joedf commented on src/SoupRot.ahk:
```diff
@@ -0,0 +1,159 @@
+		outData:=strGet(&outData,"utf-8")
+		return bytes
+	}
+	strPutVar(string,byRef var,encoding){
```
Again prefix / "self-contain" / or move it into the functions if possible.

joedf reviewed (commented):
Thanks for this! Awesome to see another addition! I made a few comments to make it more "compliant" with the rest of the library. Also, don't forget the readme: add it under `En-/Decryption (Custom)` or if more appropriate add `En-/Decoding (Custom)` I can attempt to make the changes if you do not wish to do so.

Masonjar13 committed:
Example Update

Made the output more clear.

Masonjar13 committed:
libcrypt compatibility

Removed base64 functions to use libcrypt integrated base64 functions
Altered example for clarity

Masonjar13 committed:
Naming convention update

Updated naming convention for class and dependencies

Masonjar13 commented:
I think I got all the changes, besides updating the version and running the make file. Did I miss anything?

Edit: yes the readme. I wasn't sure where to put it. I'm back and forth if this is considered a cipher or encryption. Seeing as you have the rot functions under encryption, I may as well put it there, yes?

joedf commented:
> I think I got all the changes, besides updating the version and running the make file. Did I miss anything?

Everything looks good! ready to merge to as-is, just let me know when. Or you probably have access too. 👍 

> Edit: yes the readme. I wasn't sure where to put it. I'm back and forth if this is considered a cipher or encryption. Seeing as you have the rot functions under encryption, I may as well put it there, yes?

I guess so, not sure myself. 😅  Agreed.

Masonjar13 committed:
Added SoupRot
